README
-------------------------------------------------------------------------------
                 GBAconv 1.00 (c) by Frederic Cambus 2002-2006
-------------------------------------------------------------------------------



LICENSE:
--------

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA



DESCRIPTION:
------------

GBAconv is a set of tools aimed at GameBoy Advance developers, allowing
to convert WAV sounds and PCX images into C arrays to be included in
GBA programs.

It is composed of two programs:

pcx2gba: exports a PCX 8-bpp image to 16-bit values C arrays
         (palette + image data (2 pixels per value))
wav2gba: exports a WAV 8-bit mono sound to 8-bit signed values array.

There won't be any other releases of GBAconv, as I've no interest for
the GBA platform anymore.



COMPILING:
----------

Simply type "make" to compile the programs.

They have been successfully compiled on:

- 32-bit x86 Architecture (Ubuntu Linux 5.10, FreeBSD 5.4)



USAGE:
------

pcx2gba input.pcx output.inc array_name (Input File must be 8-bit PCX)
wav2gba input.wav output.inc array_name (Input File must be 8-bit, MONO)



DISTRIBUTION:
-------------

If you want to add a GBAconv download link on your page, please use the
following sourceforge download link:

For version 1.00:
http://prdownloads.sourceforge.net/gbaconv/gbaconv-1.00-src.tar.gz?download



CONTACT:
--------

Project Homepage: http://gbaconv.sourceforge.net

For bug report, feedback, or any other reasons, write at the following 
address (replace AT by @ and DOT by .):
Frederic Cambus <fcambus AT users DOT sourceforge DOT net>


ChangeLog
1.00 (2006-01-13)

- Initial release


wav2gba.c
/*****************************************************************************/
/*                                                                           */
/* GBAconv 1.00 (c) by Frederic Cambus 2002-2006                             */
/* http://gbaconv.sourceforge.net                                            */
/*                                                                           */
/* WAV to GBA Converter                                                      */
/*                                                                           */
/* Created:      2002/12/10                                                  */
/* Last Updated: 2006/01/13                                                  */
/*                                                                           */
/* This program is free software; you can redistribute it and/or modify      */
/* it under the terms of the GNU General Public License as published by      */
/* the Free Software Foundation; either version 2 of the License, or         */
/* (at your option) any later version.                                       */
/*                                                                           */
/* This program is distributed in the hope that it will be useful,           */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of            */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             */
/* GNU General Public License for more details.                              */
/*                                                                           */
/* You should have received a copy of the GNU General Public License         */
/* along with this program; if not, write to the Free Software               */
/* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */
/*                                                                           */
/*****************************************************************************/

#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

FILE *input_file;
char *input_file_buffer;
int input_file_size;
struct stat input_file_stat;

FILE *output_file;

struct wave_header
{
   char chunk_ID[4];
   unsigned int chunk_size;
   char format[4];
   char fmt_chunk[4];
   unsigned int fmt_chunk_size;
   unsigned short int audio_format;
   unsigned short int channels;
   unsigned int sample_rate;
   unsigned int byte_rate;
   unsigned short int block_align;
   unsigned short int bits_per_sample;
   char data_chunk[4];
   unsigned int data_chunk_size;
} wave_header;

int loop;



/*****************************************************************************/
/* MAIN                                                                      */
/*****************************************************************************/

int main (int argc, char *argv[])
{
   printf("-------------------------------------------------------------------------------\n");
   printf("      WAV to GBA Converter - GBAconv 1.00 (c) by Frederic Cambus 2002-2006\n");
   printf("-------------------------------------------------------------------------------\n\n");

   if (argc!=4)
   {
      printf("USAGE: wav2gba input.wav output.inc array_name (Input File must be 8-bit, MONO)\n\n");
      exit(0);
   }



/*****************************************************************************/
/* LOAD INPUT FILE                                                           */
/*****************************************************************************/

   stat (argv[1], &input_file_stat);
   input_file_size=input_file_stat.st_size;

   input_file_buffer=malloc(input_file_size);

   if (input_file_buffer==NULL)
   {
      printf("ERROR: Cannot allocate memory\n\n");
      exit(-1);
   }

   input_file=fopen(argv[1],"rb");
   if (input_file==NULL)
   {
      printf("ERROR: Cannot open file %s\n\n",argv[1]);
      exit(-1);
   }

   fread(input_file_buffer,input_file_size,1,input_file);
   fclose(input_file);



/*****************************************************************************/
/* CHECK THAT THE FILE IS A VALID 8-bit MONO WAV                             */
/*****************************************************************************/

   memcpy(&wave_header,input_file_buffer,44);

   if (wave_header.channels!=1)
   {
      printf("ERROR: Input File is not MONO\n\n");
      exit(-1);
   }

   if (wave_header.bits_per_sample!=8)
   {
      printf("ERROR: Input File is not 8-bit\n\n");
      exit(-1);
   }



/****************************************************************************/
/* CREATE OUTPUT FILE                                                        */
/*****************************************************************************/

   output_file=fopen(argv[2],"w");
   if (output_file==NULL)
   {
      printf("ERROR: Cannot create file %s\n\n",argv[2]);
      exit(-1);
   }

   printf("INPUT  FILE: %s (8-bit, MONO, %i Hz)\n",argv[1],wave_header.sample_rate);
   printf("OUTPUT FILE: %s\n\n",argv[2]);

   fprintf(output_file,"const s8 %s[] = {\n", argv[3]);

   for (loop=0;loop<input_file_size;loop++)
   {
      fprintf(output_file,"0x%x,",input_file_buffer[44+loop]+128);
   }
   fseek(output_file,ftell(output_file)-1,0);
   fprintf(output_file,"};\n");



/*****************************************************************************/
/* TERMINATE PROGRAM                                                         */
/*****************************************************************************/

   fclose(output_file);
   free(input_file_buffer);

   printf("Successfully created file %s\n\n",argv[2]);

   return (0);
}


pcx2gba.c
/*****************************************************************************/
/*                                                                           */
/* GBAconv 1.00 (c) by Frederic Cambus 2002-2006                             */
/* http://gbaconv.sourceforge.net                                            */
/*                                                                           */
/* PCX to GBA Converter                                                      */
/*                                                                           */
/* Created:      2002/12/09                                                  */
/* Last Updated: 2006/01/13                                                  */
/*                                                                           */
/* This program is free software; you can redistribute it and/or modify      */
/* it under the terms of the GNU General Public License as published by      */
/* the Free Software Foundation; either version 2 of the License, or         */
/* (at your option) any later version.                                       */
/*                                                                           */
/* This program is distributed in the hope that it will be useful,           */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of            */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             */
/* GNU General Public License for more details.                              */
/*                                                                           */
/* You should have received a copy of the GNU General Public License         */
/* along with this program; if not, write to the Free Software               */
/* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */
/*                                                                           */
/*****************************************************************************/

#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

FILE *input_file;
unsigned char *input_file_buffer;
int input_file_size;
struct stat input_file_stat;

FILE *output_file;

struct pcx_header
{
   char ID;
   char version;
   char encoding;
   char bits_per_pixel;
   short int x_min;
   short int y_min;
   short int x_max;
   short int y_max;
   short x_resolution;
   short y_resolution;
   char palette[48];
   char reserved;
   char number_of_bit_planes;
   short bytes_per_line;
   short palette_type;
   short x_screen_size;
   short y_screen_size;
   char filler[54];
} pcx_header;

unsigned char pcx_image_palette[768];
unsigned char *pcx_buffer;
int pcx_buffer_size;

int loop;

int current_byte;
int offset;
int run_count;
int run_position;



/*****************************************************************************/
/* MAIN                                                                      */
/*****************************************************************************/

int main (int argc, char *argv[])
{
   printf("-------------------------------------------------------------------------------\n");
   printf("      PCX to GBA Converter - GBAconv 1.00 (c) by Frederic Cambus 2002-2006\n");
   printf("-------------------------------------------------------------------------------\n\n");

   if (argc!=4)
   {
      printf("USAGE: pcx2gba input.pcx output.inc array_name (Input File must be 8-bpp PCX)\n\n");
      exit(0);
   }



/*****************************************************************************/
/* LOAD INPUT FILE                                                           */
/*****************************************************************************/

   stat (argv[1], &input_file_stat);
   input_file_size=input_file_stat.st_size;

   input_file_buffer=malloc(input_file_size);

   if (input_file_buffer==NULL)
   {
      printf("ERROR: Cannot allocate memory\n\n");
      exit(-1);
   }

   input_file=fopen(argv[1],"rb");
   if (input_file==NULL)
   {
      printf("ERROR: Cannot open file %s\n\n",argv[1]);
      exit(-1);
   }

   fread(input_file_buffer,input_file_size,1,input_file);
   fclose(input_file);



/*****************************************************************************/
/* CHECK THAT THE FILE IS A VALID 8-bpp PCX                                  */
/*****************************************************************************/

   memcpy(&pcx_header,input_file_buffer,128);

   if (pcx_header.bits_per_pixel!=8)
   {
      printf("ERROR: Input File is not 8-bpp\n\n");
      exit(-1);
   }



/*****************************************************************************/
/* UNCOMPRESS RLE ENCODED PCX INPUT FILE                                     */
/*****************************************************************************/

   pcx_buffer_size=(pcx_header.x_max+1)*(pcx_header.y_max+1);
   pcx_buffer=malloc(pcx_buffer_size);

   while (loop<input_file_size-768-128)
   {
      current_byte=input_file_buffer[loop+128];

      if (current_byte>192)
      {
         run_count=current_byte-192;

         for (run_position=0;run_position<run_count;run_position++)
         {
            pcx_buffer[offset+run_position]=input_file_buffer[loop+128+1];
         }
         offset+=run_count;
         loop+=2;
      }
      else
      {
         pcx_buffer[offset]=current_byte;
         offset++;
         loop++;
      }
   }

   for (loop=0;loop<768;loop++)
   {
      pcx_image_palette[loop]=(input_file_buffer[input_file_size-768+loop]/8);
   }



/*****************************************************************************/
/* CREATE OUTPUT FILE                                                        */
/*****************************************************************************/

   output_file=fopen(argv[2],"w");
   if (output_file==NULL)
   {
      printf("ERROR: Cannot create file %s\n\n",argv[2]);
      exit(-1);
   }

   printf("INPUT  FILE: %s (%ix%ix%i-bpp)\n",argv[1],pcx_header.x_max+1,pcx_header.y_max+1,pcx_header.bits_per_pixel);
   printf("OUTPUT FILE: %s\n\n",argv[2]);

   fprintf(output_file,"const u16 %s_palette[] = {\n", argv[3]);

   for (loop=0;loop<256;loop++)
   {
      fprintf(output_file,"0x%x,",(pcx_image_palette[loop*3] | pcx_image_palette[(loop*3)+1]<<5 | pcx_image_palette[(loop*3)+2]<<10));
   }

   fseek(output_file,ftell(output_file)-1,0);
   fprintf(output_file,"};\n\n");

   fprintf(output_file,"const u16 %s[] = {\n", argv[3]);

   for (loop=0;loop<pcx_buffer_size/2;loop++)
   {
      fprintf(output_file,"0x%x,",(pcx_buffer[loop*2] | pcx_buffer[(loop*2)+1]<<8));
   }

   fseek(output_file,ftell(output_file)-1,0);
   fprintf(output_file,"};\n");



/*****************************************************************************/
/* TERMINATE PROGRAM                                                         */
/*****************************************************************************/

   fclose(output_file);
   free(input_file_buffer);

   printf("Successfully created file %s\n\n",argv[2]);

   return (0);
}
